Core.Collectors/Web/RateLimiter.cs

```diff
@@ -185,19 +183,37 @@ public static string GetRateLimitResource(HttpResponseHeaders responseHeaders)
 
         public async Task WaitIfNeededAsync(IAuthentication authentication, string resource = null)
         {
-            TimeSpan elapsedSinceLastLookup = this.DateTimeSystem.UtcNow - this.cacheDateUtc;
-            if (this.cachedResult == null || elapsedSinceLastLookup >= this.cacheInvalidationFrequency)
+            RateLimitTableEntity record = null;
+
+            string rowKey = RateLimitTableEntity.GetRowKey(this.OrganizationId, resource);
+            bool lookup = false;
+            if (this.cachedResultMap.TryGetValue(RateLimitTableEntity.GetRowKey(this.OrganizationId, resource), out Tuple<RateLimitTableEntity, DateTime> recordWithDate))
+            {
+                DateTime cacheDateUtc = recordWithDate.Item2;
+                TimeSpan elapsedSinceLastLookup = this.DateTimeSystem.UtcNow - cacheDateUtc;
+                record = recordWithDate.Item1;
+                if (!record.RowKey.Equals(rowKey) || elapsedSinceLastLookup >= this.cacheInvalidationFrequency)
+                {
+                    lookup = true;
+                }
+            }
+            else
+            {
+                lookup = true;
+            }
+
+            if (lookup)
             {
-                this.cachedResult = await this.rateLimiterCache.RetrieveAsync(new RateLimitTableEntity(authentication.Identity, this.OrganizationId, this.OrganizationName, resource)).ConfigureAwait(false);
-                this.cacheDateUtc = this.DateTimeSystem.UtcNow;
+                record = await this.rateLimiterCache.RetrieveAsync(new RateLimitTableEntity(authentication.Identity, this.OrganizationId, this.OrganizationName, resource)).ConfigureAwait(false);
+                this.cachedResultMap[rowKey] = Tuple.Create(record, this.DateTimeSystem.UtcNow);
```

Comment:
You're saving to the cache even when the retrieved record is `null`.
Leads to NPE in line 195 above: `if (!record.RowKey.Equals(rowKey) ...`